File: QuantConnect/Strategy_26.cs


# Strategy_SupportResistance.py

from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import PivotPoints, RollingWindow
from datetime import timedelta


class SupportResistance(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2015, 1, 1)
        self.SetEndDate(2020, 1, 1)
        self.SetCash(100000)
        self.AddEquity("SPY", Resolution.Daily)
        
        # Create indicators
        self.pp = PivotPoints("PP", 14)
        
        # Rolling window for price history
        self.window = RollingWindow[TradeBar](2)
        
        # Initialize variables
        self.entryPrice = None
        self.stopPrice = None
        self.exitPrice = None
    
    def OnData(self, data):
        
        if not (data.ContainsKey("SPY") and data["SPY"].Close):
            return
        
        # Add new data to rolling window
        self.window.Add(data["SPY"])
        
        # Wait for enough data to calculate pivot points
        if not all(map(lambda x: x.IsReady, [self.pp])):
            return
        
        # If we have a position and the price has reached the pivot point, exit the position
        if self.Portfolio.Invested:
            if (self.entryPrice < self.pp.Support2 and data["SPY"].Close > self.pp.Support2) or \
                    (self.entryPrice > self.pp.Resistance2 and data["SPY"].Close < self.pp.Resistance2):
                self.Liquidate()
                return
            
            if (self.entryPrice < self.pp.Support1 and data["SPY"].Close > self.exitPrice) or \
                    (self.entryPrice > self.pp.Resistance1 and data["SPY"].Close < self.exitPrice):
                return
            
            if data["SPY"].Close > max(self.exitPrice, max(self.window[0].High, 
                                                            max(self.window[0].Low + ATR("ATR", 14).Current.Value,
                                                                max(self.window[1].High - ATR("ATR", 14).Current.Value,
                                                                    max(self.pp.Resistance1 + ATR("ATR", 14).Current.Value,
                                                                        max(self.pp.Resistance2 + ATR("ATR", 14).Current.Value,
                                                                            max(self.pp.Resistance3 + ATR("ATR", 14).Current.Value,
                                                                                max(self.pp.Resistance4 + ATR("ATR", 14).Current.Value,
                                                                                    max(self.pp.Resistance5 + ATR("ATR", 14).Current.Value,
                                                                                        -float('inf')))))))))):
                return
            
            else:
                stopLoss = StopMarketOrder("StopLossOrder", -self.Portfolio[self.Symbol].Quantity, 
                                           min(data["SPY"].Low - ATR("ATR", 14).Current.Value, 
                                               min(min(self.window[0].Low,self.window[1].Low))))
                
                stopLossTag = "{}-{}".format(stopLoss.TypeName.lower(), stopLoss.Id)
                if not stopLossTag in [x.Tag for x in list(self.Transactions.GetOpenOrders())]:
                    submitResult = self.SubmitOrder(stopLoss)
                    stopLossTicket = submitResult.OrderTicket
                    
                    if stopLossTicket is not None:
                        stopLossTicket.UpdateTag(stopLossTag)

                
            
    
    def OnOrderEvent(self, orderEvent):
        
    def SetStopAndExitPrices(self):
        
            
    def RebalancePortfolio(self):

File: QuantConnect/Strategy_26.py


# Strategy_SupportResistance.py

from QuantConnect.Data.Market import TradeBar
from QuantConnect.Indicators import PivotPoints, ExponentialMovingAverage, AverageTrueRange
from datetime import timedelta


class SupportResistance(QCAlgorithm):

    def Initialize(self):
        self.SetStartDate(2015, 1, 1)  
        self.SetEndDate(2020, 12, 31)  
        self.SetCash(100000)  
        
        self.symbol = self.AddEquity("SPY", Resolution.Daily).Symbol
        
        # Indicators
        self.pivot = PivotPoints("SPY")
        self.ema = ExponentialMovingAverage(30)
        self.atr = AverageTrueRange(14)
        
        # Set warmup period
        self.SetWarmUp(timedelta(days=30))
        
    def OnData(self, data: TradeBar):
        
        if not (self.pivot.IsReady and self.ema.IsReady and self.atr.IsReady):
            return
        
        # Check if price is at pivot point
        if data.Close == self.pivot.Resistance1:
            # Buy on breakout above pivot point with ATR stop loss
            stop_price = data.Close - (self.atr.Current.Value * 1)
            self.SetStopLoss(self.symbol, stop_price)
            self.MarketOrder(self.symbol, 100)
            
        elif data.Close == self.pivot.Support1:
            # Short on breakdown below pivot point with ATR stop loss
            stop_price = data.Close + (self.atr.Current.Value * 1)
            self.SetStopLoss(self.symbol, stop_price)
            self.MarketOrder(self.symbol, -100)
            
        
    def OnOrderEvent(self, orderEvent):
        
        order = self.Transactions.GetOrderById(orderEvent.OrderId)
        
        if order.Status != OrderStatus.Filled:
            return
        
        # Exit on opposite signal or trailing stop
        position = [x for x in self.Portfolio if x.Key == order.Symbol][0].Value
        if position.Quantity > 0:
            exit_price = min(self.pivot.Support2, max(order.Price - (self.atr.Current.Value * 2), 
                                                      position.AveragePrice * (1 - 0.02)))
            if exit_price < data.Close:
                return
                
            limit_order = self.LimitOrder(order.Symbol, -position.Quantity,
                                          exit_price,
                                          "Exit Buy")
            
    
            
                
        
        
        
        
        
        
        